airflow/provider.yaml.schema.json

```diff
@@ -17,6 +17,11 @@
         "type": "string"
       }
     },
+    "provider-package": {
```

Comment:
@ashb @mik-laj -> this is also for preparation to be able to use 3rd-party providers. I've added provider-package to the provider.yaml to avoid repetition and made all modules/classes relative. This way it is really safe - because we will not be able to point accidentally to a class outside of our package.

I also added a check if provider package is properly named bu also if it starts with "airflow.providers". 

Currently we only support airflow.providers anyway. 

It might be good to discuss now and agree future conventions for those 3rd-parties eventually

1) are we allowing any package to contain provider?
If not:
2a) should all providers live in airflow.providers
2b) should they live in airflow.contrib.providers (or similar)
if 2a) and 2b) - > should we allow the packages to be named not following the convention :

apache-airflow-providers-<PROVIDER> (with '.' replaced by -). 

OR maybe some other approach? WDYT?

WDYT?

Reply:
I will also split this into two PRs - one adding "provider-package" and one adding connections, to keep it separate.

Also after looking at how fast and easy it is to extract all the "provider" features (extra links, connections ets) and keep in  the  provider.yaml,  I think maybe ewe should bring the discussion about conventions for 3rd-party providers to the devllst?  This was non-goal for AIP-8 and Airflow 2.0, but it seems we are SOOOO close to it that it's worth doing the extra effort and  discuss and vote the 3rd-party provider conventions showing examples from our existing providers.

The nice thing about it that we can provide some examples and clear instructions on "how to add a new 3rrd-party provider" and even a way to generate a really nice documentation with examples. Seems that we already pretty much have what we need for that very soon, it just needs agreement on the devlist and we still have time to discuss it before 2.0.

WDYT?

Reply:
As discussed with @mik-laj -> I will go back to absolute imports and change it to verify if they are the same as provider-package. It will be easier to maintain it.

Reply:
>         are we allowing any package to contain provider?

Yes, I firmly believe this should be the case. By not having third party packages live under the `airflow.` namespace, it becomes much much easier for Us and hopefully users to know what is an official provider, and thus supported by us vs not.

I think the best approach would be something like a setuptools entrypoint.

That way _any_ package can register a provider, by adding something like this to it's setup.cfg:

```ini
[options.entry_points]
apache-airflow-provider=
    x=my_company.provider.x:register_provider
```

And then in `my_company/provider/x.py` we could have something like this:

```python
def get_provider_info():
    return {
        description: "X"
    }
```

We don't need to include the package name, nor the version, as those are available from Entrypoint already.

In short: Python already has this mechanism that is well used. Let's not re-invent something different.

Reply:
Thanks! 

TL;DR; I think entrypoint idea for runtime optimization is really cool, happy to incorporate it as long as we keep "developer" experience intact (or even optimized in this case as well).

> Yes, I firmly believe this should be the case. By not having third party packages live under the airflow. namespace, it becomes much much easier for Us and hopefully users to know what is an official provider, and thus supported by us vs not.

1. Do you think we should prreprevent or enforce 3rd parties from using airflow.* ? I do not think there is a way of doing it unless we sign at least some of the package cryptographically with our own case? Or is it just convention you propose here? That other packages SHOULD NOT use "airflow" as package by convention? 

> That way any package can register a provider, by adding something like this to it's setup.cfg:

[options.entry_points]
apache-airflow-provider=
    x=my_company.provider.x:register_provider

That's a cool idea, especially that rather than returning a hard-coded dictionary, we can return the very same `provider.yaml` file, as a dictionary. That will make it super-easy for anyone writing their own provider as they will have the .json.schema that will define the expected structure of the file and it will be very easy to explain it. Very standard, supported by most IDEs with auto-completion and documentation (I used it to add the current files) and it can be used to verify such provider information during provider development.

This might help to speed up the runtime discovery - One of the problems with `walk_packages` is that it is a bit slow. Sp having to use it at runtime is good. And I am happy to change the mechanism for runtime for that, as long (as discussed on slack) as we solve the problem of local provider development. We need to still support testing and developing airflow providers directly from sources in checked out airflow sources. Having to install each provider as a package in order to test it sounds like a terrible idea for people who develop providers (which are most of the casual contributors to airflow). 

If for runtime we can switch to entry_points, that makes it possible also to optimize the - directly-from-sources case. This is then much easier because we can rely on packages being in the right folder and we can use simple directory walking to find all provider files, which makes it considerably faster. Paired with the runtime "entrypoint' discovery I think we can have a very optimized solution keeping all the nice properties we have now: keeping provider "meta-data" in one place, with validated json.schema yaml file in "airflow" providers. But if someone wants, they can still keep them as dict objects or json in 3rd-party providers as long as they follow the schema. I don't think we need to force anyone to yaml for that even if it is super-convenient to us and we use it also for document generation.


Unless you have other, better idea to keep the developer experience intact.

Reply:
Hey @ashb - implemented all that in #12512  and this one is only left for connections. we can discuss it separately.